Packages/Blockchain/Sources/Ethereum/EthereumFeeService.swift

```diff
@@ -114,20 +115,104 @@ extension EthereumService: ChainFeeCalculateable {
         }
     }
 
-    public func feeRates() async throws -> [FeeRate] { fatalError("not implemented") }
+    public func feeRates() async throws -> [FeeRate] {
+        let (baseFee, priorityFees) = try await getBasePriorityFees(
+            blocks: Self.historyBlocks,
+            config: GemstoneConfig.shared.config(for: chain)
+        )
+        var rates: [FeeRate] = []
+        for priority in FeePriority.allCases {
+            if let priorityFee = priorityFees[priority] {
+                let feeRate = FeeRate(
+                    priority: priority,
+                    gasPriceType: .eip1559(
+                        gasPrice: baseFee,
+                        minerFee: priorityFee
+                    )
+                )
+                rates.append(feeRate)
+            }
+        }
+        return rates
+    }
+
+    func getBasePriorityFees(blocks: Int, config: EvmChainConfig) async throws -> (base: BigInt, priority: [FeePriority: BigInt]) {
+        let feeHistory = try await provider
+            .request(.feeHistory(blocks: blocks, rewardPercentiles: config.rewardsPercentiles.list))
+            .map(as: JSONRPCResponse<EthereumFeeHistory>.self).result
+
+        guard
+            let baseFeeHex = feeHistory.baseFeePerGas.last,
+            let baseFeePerGas = try? BigInt.fromHex(baseFeeHex)
+        else {
+            throw AnyError("Unable to retrieve base fee from history")
+        }
+        let priorityFees = calculatePriorityFees(
+            rewards: feeHistory.reward,
+            config: config
+        )
+        return (baseFeePerGas, priorityFees)
+    }
+
+    func calculatePriorityFees(
+        rewards: [[String]],
+        config: EvmChainConfig
+    ) -> [FeePriority: BigInt] {
+
+        let priorityFeesPerPercentile: [Int: [BigInt]] = {
```

Comment:
```suggestion
        let priorityFeesPerPercentile: [Int: [BigInt]] = Dictionary(
    uniqueKeysWithValues: config.rewardsPercentiles.list.map { percentile in
        (percentile, rewards.compactMap { rewardsPerBlock in
            rewardsPerBlock.enumerated()
                .filter { config.rewardsPercentiles.list[$0.offset] == percentile }
                .compactMap { try? BigInt.fromHex($0.element) }
        }.flatMap { $0 })
    }
)
```

can you test this one?

Reply:
but without passing `config`